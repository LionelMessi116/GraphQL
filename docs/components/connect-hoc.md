### About

A HOC alternative implementation to the [Connect](Connect) component.

### Props

| Name      | Type                                                         | Default   | Description                                                  |
| --------- | ------------------------------------------------------------ | --------- | ------------------------------------------------------------ |
| query     | [Query](../types/mutation-query.md)?                         | undefined | The query you want connected to your component.              |
| mutations | Object { [string]: [Mutation](../types/mutation-query.md) }? | undefined | The mutation/mutations you want connected to your component. |

### Example

```jsx
import { ConnectHOC, createMutation, createQuery } from 'urql';
import { TodoList } from './components';

// ...
const connectArgs = {
  query: createQuery(QueryTodos),
  mutations: {
    addTodos: createMutation(AddTodos),
  },
};

// ...
return ConnectHOC(connectArgs)(TodoList);
```
