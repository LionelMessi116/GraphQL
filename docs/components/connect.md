### About

Connect is a ReactJS component that is used to execute queries and mutations and render child components with the results, using a render prop.

### Props

| Name      | Type                                                         | Default   | Description                                                        |
| --------- | ------------------------------------------------------------ | --------- | ------------------------------------------------------------------ |
| query     | [Query](../types/mutation-query.md)?                         | undefined | The query you want connected to your component                     |
| mutations | Object { [string]: [Mutation](../types/mutation-query.md) }? | undefined | The mutation/mutations you want connected to your component        |
| children  | ([ChildArgs\<T\>](../types/child-args.md)) => ReactNode      | undefined | A child function to accept the connected state and render elements |

### Example

```jsx
import { Connect, createQuery, createMutation } from 'urql';

// ...
<Connect
  query={createQuery(GetTodos)}
  mutations={{
    addTodo: createMutation(AddTodo)
  }}
  children={({ fetching, data }) => {
    return fetching ? <Loading/> : <List data={data.todos}>
  }}
/>
```
