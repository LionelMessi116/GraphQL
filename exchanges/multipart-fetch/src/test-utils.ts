import {
  GraphQLRequest,
  OperationContext,
  Operation,
  makeOperation,
} from '@urql/core';
import gql from 'graphql-tag';

const context: OperationContext = {
  fetchOptions: {
    method: 'POST',
  },
  requestPolicy: 'cache-first',
  url: 'http://localhost:3000/graphql',
};

const queryGql: GraphQLRequest = {
  key: 2,
  query: gql`
    query getUser($name: String) {
      user(name: $name) {
        id
        firstName
        lastName
      }
    }
  `,
  variables: {
    name: 'Clara',
  },
};

const obj = { hello: 'world' };
const file = new File(
  [new Blob([JSON.stringify(obj, null, 2)], { type: 'application/json' })],
  'index.ts'
);
const files = [file, file];

const upload: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation uploadProfilePicture($picture: File) {
      uploadProfilePicture(picture: $picture) {
        location
      }
    }
  `,
  variables: {
    picture: file,
  },
};

const uploads: GraphQLRequest = {
  key: 3,
  query: gql`
    mutation uploadProfilePictures($pictures: [File]) {
      uploadProfilePicture(pictures: $pictures) {
        location
      }
    }
  `,
  variables: {
    picture: files,
  },
};

export const uploadOperation: Operation = makeOperation(
  'mutation',
  upload,
  context
);

export const multipleUploadOperation: Operation = makeOperation(
  'mutation',
  uploads,
  context
);

export const queryOperation: Operation = makeOperation(
  'query',
  queryGql,
  context
);
