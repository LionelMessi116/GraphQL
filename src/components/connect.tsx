import React, { ReactNode } from 'react';
import { Client } from '../lib/client';
import { ChildArgs, Mutation, Query, Subscription } from '../types';
import { UrqlClient } from './client';
import { Consumer } from './context';

/** Props for the [Connect]{@link Connect} component. */
export interface ConnectProps<T> {
  /** A function which receives values from the URQL client. */
  children: (props: ChildArgs<T>) => ReactNode;
  /** The GraphQL query to fetch */
  query?: Query;
  /** A collection of GrahpQL mutation queries */
  mutations?: { [type in keyof T]: Mutation };
  /** An array of GrahpQL subscription queries */
  subscription?: Subscription;
  /** An updator function for merging subscription responses */
  updateSubscription?: <Data, Update>(prev: Data, next: Update) => Data;
}

/** Component for connecting to the urql client for executing queries, mutations and returning the result to child components. */
export function Connect<T>(props: ConnectProps<T>) {
  return (
    <Consumer>
      {client => <UrqlClient {...props} client={client} />}
    </Consumer>
  );
}
