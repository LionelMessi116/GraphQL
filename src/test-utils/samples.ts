import {
  ExchangeResult,
  ExecutionResult,
  Mutation,
  Operation,
  Query,
  Subscription,
} from '../types';

const context = {
  fetchOptions: {
    method: 'POST',
  },
  url: 'http://localhost:3000/graphql',
};

export const queryGql: Query = {
  query: `query getUser($name: String){
    user(name: $name) {
      id
      firstName
      lastName
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const mutationGql: Mutation = {
  query: `mutation AddUser($name: String){
    addUser(name: $name) {
      name
    }
  }`,
  variables: {
    name: 'Clara',
  },
};

export const subscriptionGql: Subscription = {
  query: `subscription subscribeToUser($user: String){
    user(user: $user) {
      name
    }
  }
  `,
  variables: {
    user: 'colin',
  },
};

export const queryOperation: Operation = {
  key: '2',
  operationName: 'query',
  context,
  ...queryGql,
};

export const mutationOperation: Operation = {
  key: JSON.stringify(mutationGql),
  operationName: 'mutation',
  context,
  ...mutationGql,
};

export const subscriptionOperation: Operation = {
  key: JSON.stringify(subscriptionGql),
  operationName: 'subscription',
  context,
  ...subscriptionGql,
};

export const queryResponse: ExchangeResult = {
  operation: queryOperation,
  data: {
    user: {
      name: 'Clive',
    },
  },
};

export const mutationResponse: ExchangeResult = {
  operation: mutationOperation,
  data: {},
};

export const subscriptionResult: ExecutionResult = {
  data: {}
};
