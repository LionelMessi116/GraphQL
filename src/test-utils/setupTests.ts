import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

// Setup enzyme adapter
configure({ adapter: new Adapter() });

// Setup globals
const g = global as any;

g.AbortController = undefined;
g.fetch = jest.fn();

process.on('unhandledRejection', error => {
  throw error;
});
