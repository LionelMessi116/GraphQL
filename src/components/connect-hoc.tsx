import React, { ComponentType } from 'react';
import { Connect, ConnectProps } from '../components/connect';
import { ChildArgs } from '../types';

/** Props for the [ConnectHOC]{@link ConnectHOC} component. */
export interface ConnectHOCProps<T> {
  /** The GraphQL query to fetch. */
  query?: ConnectProps<T>['query'];
  /** A collection of GrahpQL mutation queries. */
  mutations?: ConnectProps<T>['mutations'];
}

/** A HOC alternative implementation to the [Connect]{@link Connect} component. */
export const ConnectHOC = function<T>(opts?: ConnectHOCProps<T>) {
  return (Comp: ComponentType<ChildArgs<T>>) => {
    const children = (args: ChildArgs<T>) => <Comp {...args} />;
    children.displayName = Comp.displayName || Comp.name || 'Component';

    const connected = () => <Connect {...opts} children={children} />;
    connected.displayName = `Connect(${children.displayName})`;

    return connected;
  };
};
