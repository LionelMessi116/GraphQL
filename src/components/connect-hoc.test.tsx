jest.mock('./connect');
import { mount } from 'enzyme';
import React from 'react';
import { Connect as ConnectMock } from './connect';
import { ConnectHOC } from './connect-hoc';
import { mutationGql, queryGql } from '../test-utils';

const Connect = ConnectMock as jest.Mock;
Connect.mockReturnValue(<div />);

const connectOpts = {
  mutations: { test: mutationGql },
  query: queryGql,
};

const TargetComponent = function() {
  return <div />;
};

beforeEach(() => {
  Connect.mockClear();
});

it('passes snapshot', () => {
  const Connected = ConnectHOC(connectOpts)(TargetComponent);
  const wrapper = mount(<Connected />);

  expect(wrapper).toMatchSnapshot();
});

it('Names connected component Connect(<ChildName>)', () => {
  const Connected = ConnectHOC(connectOpts)(TargetComponent);
  const wrapper = mount(<Connected />);

  expect(wrapper.find('Connect(TargetComponent)').length).toBe(1);
});

it('Calls Connect component', () => {
  const Connected = ConnectHOC(connectOpts)(TargetComponent);
  const wrapper = mount(<Connected />);

  expect(Connect).toBeCalled();
});

it('Calls Connect component with opts', () => {
  const Connected = ConnectHOC(connectOpts)(TargetComponent);
  mount(<Connected />);

  expect(Connect.mock.calls[0][0]).toEqual(
    expect.objectContaining(connectOpts)
  );
});

it('Calls Connect component with children', () => {
  const Connected = ConnectHOC(connectOpts)(TargetComponent);
  mount(<Connected />);

  expect(Connect.mock.calls[0][0].children.displayName).toEqual(
    'TargetComponent'
  );
});
