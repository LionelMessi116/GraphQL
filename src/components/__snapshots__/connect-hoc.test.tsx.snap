// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`passes snapshot 1`] = `
<Connect(TargetComponent)>
  <Connect
    mutations={
      Object {
        "test": Object {
          "query": "mutation AddUser($name: String){
    addUser(name: $name) {
      name
    }
  }",
          "variables": Object {
            "name": "Clara",
          },
        },
      }
    }
    query={
      Object {
        "query": "query getUser($name: String){
    user(name: $name) {
      id
      firstName
      lastName
    }
  }",
        "variables": Object {
          "name": "Clara",
        },
      }
    }
  >
    <div />
  </Connect>
</Connect(TargetComponent)>
`;
